### PR TITLE
SONARJAVA-1740 Propagate defered type

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java [hunk 1]
```diff
@@ -221,18 +222,27 @@ public void visitMethodInvocation(MethodInvocationTree tree) {
...
+      symbol = Symbols.unknownSymbol;
+    } else {
+      symbol = resolution.symbol();
+      returnType = resolution.type();
+    }
     mit.setSymbol(symbol);
-    JavaType returnType = resolution.type();
-    if(resolution.symbol().isMethodSymbol()) {
+    if(symbol.isMethodSymbol()) {
       MethodJavaType methodType = (MethodJavaType) resolution.type();
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java]: BLOCKER NullPointerException might be thrown as 'resolution' is nullable here [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* BLOCKER 1 blocker

Watch the comments in this conversation to review them.

File: java-frontend/src/test/java/org/sonar/java/resolve/SymbolTableTest.java [hunk 2]
```diff
@@ -1109,4 +1111,15 @@ public void inference_on_parameterized_method_with_no_arg() {
...
+
+  @Test
+  public void lookup_method_on_defered_type() throws Exception {
+    Result result = Result.createFor("InferedCascadedReturnType");
+    JavaSymbol sortKeysByValue = result.symbol("sortKeysByValue");
+    JavaSymbol reverse = result.symbol("reverse");
+    //lookup on defered type allow method resolution
+    assertThat(sortKeysByValue.usages()).hasSize(1);
+    //Lack of resolution when target type is deduced, we should be able to redo the lookup
+    assertThat(reverse.usages()).isEmpty();
```
Comment (benzonico) [inline java-frontend/src/test/java/org/sonar/java/resolve/SymbolTableTest.java]: cf https://jira.sonarsource.com/browse/SONARJAVA-1741

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java (hunk 1 above)
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java]: FP as this cannot be entered if resolution is null.

**Pull request merged**